pkg/collector/edp/collector.go

```diff
@@ -2,69 +2,192 @@ package edp
-		trace.WithAttributes(
-			attribute.String("provider", runtime.ProviderType),
-			attribute.String("shoot_id", runtime.ShootInfo.ShootName),
-		),
-	)
+	var errs []error
+
+	childCtx, span := otel.Tracer("").Start(ctx, "kmc.collect_scans_and_send_measurements", kmcotel.SpanAttributes(runtime))
 	defer span.End()
 
-	scans := make(collector.ScanMap)
+	currentTimestamp := getTimestampNow()
+
+	scans, err := c.executeScans(childCtx, previousScans, runtime)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to successfully execute one or more scans : %w", err))
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+	}
+
+	scans, EDPMeasurements, err := c.convertScansToEDPMeasurements(scans, previousScans, runtime)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to convert one or more scans to EDP measurements: %w", err))
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+	}
+
+	if len(EDPMeasurements) == 0 {
+		errs = append(errs, errNoMeasurementsSent)
+		span.RecordError(err)
+		span.SetStatus(codes.Error, errNoMeasurementsSent.Error())
+
+		return scans, errors.Join(errs...)
+	}
+
+	payload := newPayload(
+		runtime.RuntimeID,
+		runtime.SubAccountID,
+		runtime.ShootName,
+		currentTimestamp,
+		EDPMeasurements,
+	)
+
+	err = c.sendPayload(payload, runtime.SubAccountID)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to send payload to EDP: %w", err))
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+	}
+
+	return scans, errors.Join(errs...)
+}
+
+func (c *Collector) executeScans(ctx context.Context, previousScans collector.ScanMap, runtime *runtime.Info) (collector.ScanMap, error) {
+	var errs []error
+
+	currentScans := make(collector.ScanMap)
 
 	for _, s := range c.scanners {
-		// record metrics about success/failure
-		// record spans for timing
-		scan, err := s.Scan(childCtx, runtime)
+		scan, err := s.Scan(ctx, runtime)
+		success := err == nil
+		collector.RecordScan(success, string(s.ID()), *runtime)
+		// if the scanner fails during execution, attempt to get the previous scan
 		if err != nil {
-			// log errors here, but continue with other measures
-			c.logger.Error("error measuring", zap.Error(err))
-			// use previous measure
-			scan = previousScans[s.ID()]
+			errs = append(errs, fmt.Errorf("scanner with ID(%s) failed during scanning: %w", s.ID(), err))
+
+			previousScan, previousScanExists := previousScans[s.ID()]
+			// if the previous scan also doesn't exist, nothing else we can do here
+			if !previousScanExists {
+				// since even a previous scan doesn't exist, we won't be able to convert any scan to an EDP measurement, so conversion is recorded as unsuccessful
+				collector.RecordScanConversion(false, string(s.ID()), collector.EDPBackendName, *runtime)
+				errs = append(errs, fmt.Errorf("no previous scan found for scanner with ID(%s)", s.ID()))
+
+				continue
+			}
+
+			scan = previousScan
 		}
```

Comment:
how about this (remove nested ifs and makes the flow easier to understand (imho))
```
	for _, s := range c.scanners {
		scan, err := s.Scan(ctx, runtime)
		success := err == nil
		collector.RecordScan(success, string(s.ID()), *runtime)
		// if the scanner fails during execution, attempt to get the previous scan
		if success {
			currentScans[s.ID()] = scan
			continue
		}

		// we encountered an error during scanning, so we need to attempt to get the previous scan
		errs = append(errs, fmt.Errorf("scanner with ID(%s) failed during scanning: %w", s.ID(), err))

		previousScan, exists := previousScans[s.ID()]
		// if the previous scan also doesn't exist, nothing else we can do here
		if exists {
			currentScans[s.ID()] = previousScan
			continue
		}
		// since even a previous scan doesn't exist, we won't be able to convert any scan to an EDP measurement, so conversion is recorded as unsuccessful
		collector.RecordScanConversion(false, string(s.ID()), collector.EDPBackendName, *runtime)
		errs = append(errs, fmt.Errorf("no previous scan found for scanner with ID(%s)", s.ID()))

	}
```